tests/TestNoseHooverThermostat.h

```diff
@@ -0,0 +1,293 @@
+        mean_temp = (i*mean_temp + instantaneous_temperature)/(i+1);
+        double energy = KE + PE + integrator.computeHeatBathEnergy();
+#if DEBUG
+        if(i % 10 == 0) std::cout << " Time: " << std::setw(4) << time 
+                                  << " Temperature: " << std::setw(8) << std::setprecision(3) << temperature
+                                  << " Mean Temperature: " << std::setw(8) << std::setprecision(3) << mean_temp
+                                  << " Total Energy: " << std::setw(12) << std::setprecision(8) << energy << std::endl;
+#endif
+        energies[i] = energy;
+    }
+    double sum = std::accumulate(energies.begin(), energies.end(), 0.0);
+    double mean = sum / energies.size();
+    double sq_sum = std::inner_product(energies.begin(), energies.end(), energies.begin(), 0.0);
+    double std = std::sqrt(sq_sum / energies.size() - mean * mean);
+    double relative_std = std / mean;
+    //std::cout << "Relative STD of energy " << relative_std << std::endl;
+    // Check mean temperature
+    ASSERT_EQUAL_TOL(temperature, temperature, 1e-2);
+    // Check fluctuation of conserved (total bath + system) energy
+    ASSERT_EQUAL_TOL(relative_std, 0, 1e-4);
+}
+
+void testCheckpoints() {
+    double timeStep = 0.001;
+    NoseHooverIntegrator integrator(timeStep), newIntegrator(timeStep);
+    System system;
+    double mass = 1;
+    system.addParticle(mass);
+    system.addParticle(mass);
+    NonbondedForce* force = new NonbondedForce();
+    force->addParticle(0, 0.1, 1.0);
+    force->addParticle(0, 0.1, 1.0);
+    system.addForce(force);
+    double kineticEnergy = 1e6;
+    double temperature=300, collisionFrequency=1, chainLength=3, numMTS=3, numYS=3;
+    chainLength = 10;
+    integrator.addSubsystemThermostat(std::vector<int>(), std::vector<std::pair<int,int>>{{0,1}},  temperature, collisionFrequency, temperature, collisionFrequency,
+                                      chainLength, numMTS, numYS);
+    newIntegrator.addSubsystemThermostat(std::vector<int>(), std::vector<std::pair<int,int>>{{0,1}},  temperature, collisionFrequency, temperature, collisionFrequency,
+                                      chainLength, numMTS, numYS);
+    Context context(system, integrator, platform);
+    Context newContext(system, newIntegrator, platform);
+    std::vector<Vec3> positions(2);
+    positions[1] = {0.1,0.1,0.1};
+    context.setPositions(positions);
+    integrator.step(300);
+#if DEBUG
+    std::cout << std::endl << std::endl << std::endl << "writing checkpoint";
+#endif
+    std::stringstream checkpoint; 
+    context.createCheckpoint(checkpoint);
+    
+    State state = context.getState(State::Positions | State::Velocities);
+    for (size_t i=0; i<100; i++){
+        state = context.getState(State::Positions | State::Velocities);
+#if DEBUG
+        std::cout << "posvel" << state.getPositions()[0] << " " << state.getVelocities()[0] << std::endl;
+#endif
+        integrator.step(1);
+    }
```

Comment:
I don't quite understand this loop.  Could it all be simplified to

```c++
integrator.step(100)
State state = context.getState(State::Positions | State::Velocities);
```

Reply:
That's right. This is just us debugging. Will be removed.